README.md
weewx-brultech
======

Extension to the WeeWX weather system for collecting data and generating reports for the
Brultech GEM energy monitors.

## Overview

This is a extension for the WeeWX weather system for the Brultech GEM energy monitor. While an
energy monitor is very different from a weather station, WeeWX is a very flexible system and has no
trouble running the driver. The extension includes three parts:

- A driver for the GEM energy monitor. The driver is designed to poll your energy monitor at 
regular intervals, typically every 10-15 seconds, then pass the data on to the rest of the WeeWX
system.

- A database schema designed for the monitor. Because the types used by an energy monitor are so
  radically different from a weather station, this extension includes a specialty schema (called
  `gem_schema`), specifically designed for the Brultech monitors. It will be used to initialize a
  separate database from the conventional WeeWX weather database, allowing both to be run
  simultaneously. By default, it will include space for 32 channels.

- A WeeWX skin, called _SeasonsPower_, designed to display current, week, month, and year energy use. It
also includes monthly energy summaries. It can easily be customized by following the directions
in the _[WeeWX Customization Guide](http://www.weewx.com/docs/customizing.htm)_.

What follows are directions for installing the extension. It consists of five parts:
1. [Configuring your Brultech monitor](#Configuring-your-Brultech-device)
2. [Installing WeeWX](#Install-WeeWX)
3. Installing the extension, either by using the WeeWX 
[installer](#Install-the-extension-by-using-the-WeeWX-installer),
or [manually](#Installing-the-extension-manually).
4. [Select database options](#Select-database-options)
5. [Start up WeeWX](#Starting-WeeWX)

## Configuring your Brultech device
The Brultech energy monitors come with a bewildering array of modes and options. To keep things
simple, this driver makes two assumptions:

- The device will always be in server mode.
- The device driver will always poll the device.

This makes it pretty easy to configure the device:

1. Set the GEM to "server mode".
Using a browser, connect to your GEM (http://192.168.1.104, in my case) 
and go to the "Application Settings" page. 
Set it to server mode, using port 8000:

    ![Application settings](images/server_mode.png)

    Click Apply.
    
2. Then click on the "Device Management" tab, and click the "Restart" button under 
*Restart Module*.

    ![Restart Module](images/restart_module.png)
    
The monitor will reboot and be ready to receive commands from the driver.
That's it for the monitor. 

Next step is to install WeeWX.

## Install WeeWX

It is strongly recommended that you use the "setup.py" method for installing WeeWX. All of the
instructions that follow assume this.

Follow the instructions in the [setup.py install guide](http://www.weewx.com/docs/setup.htm) and
install WeeWX. When asked, specify that the `Simulator` driver be used. When you're done, WeeWX
will be installed in `/home/weewx`. 

### Deactivate weather skins
By default, when WeeWX is installed, it installs the _Seasons_ skin and activates it. You will
need to deactivate it. In the WeeWX configuration file `weewx.conf`, generally located in
`/home/weewx/weewx.conf`, set option `enable` to `false` for the _Seasons_ report. When you're
done, it will look like this:

```ini
    ...
    [[SeasonsReport]]
        # The SeasonsReport uses the 'Seasons' skin, which contains the
        # images, templates and plots for the report.
        skin = Seasons
        enable = false
        lang = en
        unit_system = us
```

Next step is to install the weewx-brultech extension itself, either by using the WeeWX installer,
or manually.

## Install the extension by using the WeeWX installer

This section describes how to install `weewx-brultech` using the WeeWX installer.

### Download and install
Download the `weewx-brultech` package and install it using the WeeWX installer:

```shell
cd /home/weewx
wget https://github.com/tkeffer/weewx-brultech/archive/master.zip
./bin/wee_extension --install=master.zip
```

### Configure
When you're done with the installation, your WeeWX configuration will be set up to use the Brultech
driver. However, you still have to configure the driver. Do this by using the `wee_config`
utility with the `--reconfigure` option:
```shell
./bin/wee_config --reconfigure
```

It will give you the opportunity to change various settings. Eventually, it will ask about options
for the Brultech driver:

```
Specify the IP address (e.g., 192.168.0.10) or hostname of the Brultech monitor.
host [192.168.1.104]: 192.168.0.7
Specify the port
port [8000]: 
Saved backup to /home/weewx/weewx.conf.20211025164234
```

In this example, we have changed the host from `192.168.1.104` to `192.168.0.7`, but accepted
the default port (`8000`).

## Installing the extension manually

This section describes how to manually configure WeeWX to use this extension.

### Download and unpack

Download the `weewx-brultech` package and unpack it

```shell
cd /home/weewx
wget https://github.com/tkeffer/weewx-brultech/archive/master.zip
tar xvf master.zip
```

### Copy files
Put the files `brultech.py` and `gem_schema.py` in the `user` subdirectory. Typically,

```shell
cd /home/weewx/weewx-brultech-master
cp bin/user/brultech.py /home/weewx/bin/user
cp bin/user/gem_schema.py /home/weewx/bin/user
```

Copy the `StandardPower` skin over:
```shell script
cp -r skins/StandardPower /home/weewx/skins
```

### Configure `weewx.conf`

This section is about manually configuring the configuration file, `weewx.conf`, usually found
in `/home/weewx/weewx.conf`.

1. __Add section `[Brultech]`__

    Insert a new stanza `[Brultech]` into your `weewx.conf` file that looks like this:

    ```ini
    [Brultech]

        # See README.md for instructions on how to configure the Brultech devices!!
    
        # Power is computed as the difference between energy records. How old a 
        # record can be and still be used for this calculation:
        stale = 1800
    
        # How often to poll the device for data
        poll_interval = 5

        # Max number of channels to emit. Limit is set by hardware (48 for GEM).
        max_channels = 32
    
        # The type of packet to use. Possible choices are GEMBin48NetTime, GEMBin48Net,
        # or GEMAscii:
        packet_type = GEMBin48NetTime
    
        # Max number of times to try an I/O operation before declaring an error
        max_tries = 3
    
        driver = user.brultech

        # The type of connection to use. It should match a section below. 
        # Right now, only 'socket' is supported.
        connection = socket
    
        # The following is for socket connections: 
        [[socket]]
            host = 192.168.1.104
            port = 8000
            timeout = 20
            # After sending a command, how long to wait before looking for a response    
            send_delay = 0.2

        [[sensor_map]]
    ```

    Be sure to set options `host` and `port` to their proper values for your network configuration.

2. __Set station type__

    Tell WeeWX to use the Brultech device driver by setting `station_type`, 
under section `[Station]`, to `Brultech`:
 
     ```ini
    [Station]

        ...

        # Set to type of station hardware. There must be a corresponding stanza
        # in this file with a 'driver' parameter indicating the driver to be used.
        station_type = Brultech
    ```

3. __Configure new binding__

    Under stanza `[DataBindings]`, add a new database binding `bt_binding` for the Brultech device.
    It should look like this: 

    ```ini
   [DataBindings]

     ...

        [[bt_binding]]
            # The database must match one of the sections in [Databases].
            database = bt_sqlite
            # The name of the table within the database
            table_name = archive
            # The manager handles aggregation of data for historical summaries
            manager = weewx.manager.DaySummaryManager
            # The schema defines the structure of the database.
            # It is *only* used when the database is created.
            schema = user.gem_schema.schema
    ```

4. __Configure new database__

    Under stanza `[Databases]`, add a new database `bt_sqlite`:
    ```ini
    [Databases]
   
    ...
   
        [[bt_sqlite]]
            database_name = weepwr.sdb
            database_type = SQLite  
    ```
   
5. __Use the new binding__ 

    Under stanza `[StdArchive]`, set the option `data_binding` to `bt_binding`. This will cause
    new data to be archived in the database dedicated to Brultech data.
    
    ```ini
    [StdArchive]

        ...
   
        data_binding = bt_binding         
   ```
   
6. __Configure and activate the _SeasonsPower_ skin__
 
   Add a subsection to `[StdReport]` for the _SeasonsPower_ skin, and activate it by
   setting `enable` to `True`:
    
   ```ini
   [StdReport]
    
       ...
    
       [[PowerReport]]
           skin = SeasonsPower
           enable = True
           data_binding = bt_binding
           [[[Units]]]
               [[[[Groups]]]]
                   group_energy2 = kilowatt_hour
   ```
   Incidentally, while you're in there, make sure that any "weather" skins have been deactivated 
   --- they won't work with the data from the energy monitor!
    
8. __Make sure the configuration service runs__
 
    Because of its many specialized types, the Brultech driver requires setting up some custom
 configurations. This is done by the service `brultech.BrultechService`. You must
 add it to the list of services to run by adding it to `data_services`. So, now
 your `[Engine]` section looks something like this: 
 
    ```ini
    [Engine]
    
       [[Services]]
           # This section specifies the services that should be run. They are
           # grouped by type, and the order of services within each group
           # determines the order in which the services will be run.
           prep_services = weewx.engine.StdTimeSynch,
           data_services = user.brultech.BrultechService
           process_services = weewx.engine.StdConvert, weewx.engine.StdCalibrate, weewx.engine.StdQC, weewx.wxservices.StdWXCalculate
           ...
    ```

## Select database options
The default schema works well for most applications. However, you may want to tune it for special
cases. 

Take a look in the schema's file, usually located in `/home/weewx/bin/user/gem_schema.py`, and you will see
a number of options. The defaults generally work well. Here are their meanings


| Option                          | Default | Meaning                                                                   |
|---------------------------------|---------|---------------------------------------------------------------------------|
| `MAX_CURRENT_CHANNELS`          | `32`    | How many current channels to include in the schema. Max for GEM is 32.    |
| `MAX_TEMPERATURE_CHANNELS`      | `8`     | How many temperature channels to include in the schema. Max for GEM is 8. |
| `MAX_PULSE_CHANNELS`            | `4`     | How many pulse (counter) channels to include. Max for GEM is 4.           |
| `INCLUDE_ACCUMULATED_ABSOLUTE`  | `True`  | Store accumulated (since device reset) absolute energy.                   |
| `INCLUDE_ACCUMULATED_POLARIZED` | `False` | Store accumulated (since device reset) polarized energy.                  |
| `INCLUDE_DELTA_ABSOLUTE`        | `True`  | Store delta absolute energy.                                              |
| `INCLUDE_DELTA_POLARIZED`       | `False` | Store delta polarized energy.                                             |
| `INCLUDE_POWER_ABSOLUTE`        | `False` | Store absolute power consumed during archive period.                      |
| `INCLUDE_POWER_POLARIZED`       | `False` | Store polarized power consumed during archive period.                     |


## Starting WeeWX
You're done installing and configuring the extension. You can start WeeWX either directly from the
command line, or start it as a daemon. Instructions are in the WeeWX User's Guide, in the section
[_Running WeeWX_](http://www.weewx.com/docs/usersguide.htm#running).

## Observation types

The driver can emit the following types

| Type               | Python type | Meaning                            |
|--------------------|-------------|------------------------------------|
| `dateTime`         | `int`       | Timestamp in unix epoch time       |
| `serial`           | `str`       | The unit serial number             |
| `unit_id`          | `int`       | Unit identification number         |
| `secs`             | `int`       | Continuous device uptime           |
| `chNN_count`       | `int`       | Pulse count in channel `NN`        |
| `chNN_volt`        | `float`     | Voltage in channel `NN`            |
| `chNN_temperature` | `float`     | Temperature in channel `NN`        |
| `chNN_amp`         | `float`     | Current (amperage) in channel `NN` |
| `chNN_P_power`     | `float`     | Average power (see below).         |
| `chNN_PD_energy2`  | `float`     | Energy (see  below)                |

NB: the channel number uses no leading zero. For example, it is `ch2_volt`, NOT `ch02_volt`.


### Energy and power

Because energy and power can be either polarized or absolute, an additional character `P` is used. In addition,
energy can mean either the accumulated energy since device reset, or just the energy over the archive interval
(see below).

    chNN_PD_energy2
    chNN_P_power

| Character | Meaning                                                                                                                                                       |
|-----------|---------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NN        | Channel number. No leading zero is used.                                                                                                                      |
| P         | Polarity , either `a` (absolute) or `p` (polarized).                                                                                                          |
| D         | If the character `d` is present (not implemented), then the type represents energy used during the archive interval. Otherwise, it is the accumulated energy. |

So, for example:
- `ch14_p_energy2` would be the accumulated polarized energy in channel 14.
- `ch2_a_energy2` would be the accumulated absolute energy in channel 2.
- `ch2_ad_energy2` would be the absolute energy used during the archive interval in channel 2.

## Accumulated vs delta energy.

### Accumulated

The Brultech energy monitors emit *accumulated* energy in units of watt-seconds. Accumulated energy is the amount of
energy that has flowed through the channel since the monitor was reset. Because a watt-second is a small amount of
energy, and it may be years since your monitor was reset, this can often be a very large number. Fortunately, the
Brultech monitors reserve five bytes for this value, so it can count up to over 300,000 kWh. A typical home might use
10,000 kWh per year, so the value is unlikely to rollover for years.

### Delta

WeeWX uses accumulated energy to calculate *delta* energy, which is the amount of energy that flowed through the channel
in an archive interval (typically 5 minutes). It is the *difference* in accumulated energy between the beginning and the
end of the interval.

### Calculations
To calculate the amount of power used over, say, a month, you can either take the difference in accumulated
energy, or you can sum the smaller delta energies for the month. Here's an example using channel 8:

| Strategy                           | Tag                          |
|------------------------------------|------------------------------|
| Calculate using accumulated energy | `$month.ch8_a_energy2.diff`  |
| Calculate using delta energy       | `$month.ch8_ad_energy2.sum`  |


### Which to use?
By default, both values are stored in the database. The skins that come with `weewx-brultech` use accumulated energy.

They each have their advantages.

#### Advantage accumulated energy

The calculation is very efficient: only two numbers need be looked up in the database. By contrast, using delta energies
requires summing all the records.

If the computer running WeeWX goes down during the time period, but the energy monitor stays up, it won't affect the
final result. This is because you only need the first and last value for the time period to calculate the total energy
used.

Of course, if the monitor goes down, you will miss energy used during that time period no matter which method you use.

#### Advantage delta energy

If you have technical problems with your monitor and must do a system reset, the counters will go to zero. This means
that if you try to calculate total energy used over a period that includes the reset, and you do the calculation by
calculating differences in accumulated energy, you will get a large negative number. By contrast, if you sum the
delta energies, you will get the correct answer.

The same thing can happen if the battery in you GEM fails. It is responsible for maintaining the count should the power
fail.


changelog
2.2.0 07-Sep-2022
* Switched energy summaries back to using differences of accumulated energy.
  This is more robust to power failure, although less robust to system reset. See the README.
* Changed default port from 8083 to 8000

2.1.0 03-Sep-2022
* Added Matthew's "Seasons" based skin. Called it "SeasonsPower".

2.0.0 05-Aug-2022
* Switch from storing accumulated energy, to delta energies
* Improved output from wee_device --info query.



util/calc_deltas.py
"""Augment a version 1 style energy database, which stores accumulated energies, with delta energies.

The WeeWX API must be in your PYTHONPATH. Generally, the following will work:

PYTHONPATH=/home/weewx/bin python3 -m calc_deltas /path/to/SQLitedatabase
"""
import argparse
import re
import sys
import time

import weedb
import weeutil.weeutil
import weewx.manager

energy2_re = re.compile(r'^ch[0-9]+(_[ap])?_energy2$')

description = """Augment older-style weewx-brultech databases, which stored only accumulated energy, to one
that stores delta energies as well. 

The accumulated energies will be left untouched. New columns with delta energies will be added. 

SQLITE ONLY!
"""

usage = """%(prog)s --help
       %(prog)s SQLITE-FILE"""

SQL_UPDATE_TEMPLATE = """
WITH archive_lag AS (SELECT dateTime as this_time,
                            LAG(dateTime) OVER (order by dateTime) as prev_time,
                            %(energy_name)s as this_val,
                            LAG(%(energy_name)s) OVER (ORDER BY dateTime) as prev_val
                     FROM archive)
UPDATE archive
SET interval       = (SELECT ifnull((this_time - prev_time) / 60, interval)
                      FROM archive_lag
                      WHERE dateTime = archive_lag.this_time),
    %(delta_name)s = (SELECT iif((this_val - prev_val)>=0, (this_val - prev_val), null) 
                      FROM archive_lag 
                      WHERE dateTime = archive_lag.this_time) 
"""


def main():
    parser = argparse.ArgumentParser(description=description,
                                     usage=usage,
                                     prog='calc_deltas')
    parser.add_argument("sqlite_file",
                        metavar="SQLITE_FILE",
                        help="Path to the SQLite weewx-brultech database")
    namespace = parser.parse_args()

    # It's possible the daily summaries have already been dropped. If so, catch the exception
    # and switch database managers.
    try:
        db_manager = weewx.manager.DaySummaryManager.open({'database_name': namespace.sqlite_file,
                                                           'driver': 'weedb.sqlite'})
    except weedb.OperationalError:
        db_manager = weewx.manager.Manager.open({'database_name': namespace.sqlite_file,
                                                 'driver': 'weedb.sqlite'})

    # Find the names of all the accumulated energy fields. These will look like (for example) 'ch5_a_energy2'
    energy_names = list(filter(lambda key: energy2_re.match(key), db_manager.obskeys))

    # Perform a natural sort before displaying
    energy_names.sort(key=weeutil.weeutil.natural_keys)
    print("Ready to augment columns:", energy_names)

    ans = weeutil.weeutil.y_or_n("Drop daily summaries and proceed y/n? ")
    if ans == 'n':
        exit("Nothing done")

    # Before dropping the daily summaries, check to see if it has already been done
    if hasattr(db_manager, 'daykeys'):
        db_manager.drop_daily()
        print("Dropped daily summaries")

    print("Adding new columns")
    t1 = time.time()

    for energy_name in energy_names:
        delta_name = energy_name.replace("_energy2", "d_energy2")

        with weedb.Transaction(db_manager.connection) as cursor:
            # Try adding the new column. If it already exists, an exception will be raised.
            # Be prepared to catch it.
            try:
                cursor.execute("ALTER TABLE archive ADD COLUMN %s INTEGER" % delta_name)
            except weedb.OperationalError:
                pass
            sql_update = SQL_UPDATE_TEMPLATE % {'energy_name': energy_name, 'delta_name': delta_name}
            cursor.execute(sql_update)

        # Give the user a sense of progress:
        print("...added", delta_name, end='\r')
        sys.stdout.flush()

    t2 = time.time()

    print("Finished adding %d new columns in %.2f seconds" % (len(energy_names), t2 - t1))


if __name__ == "__main__":
    main()


skins/SeasonsPower/power.js
/* javascript for the weewx SeasonsPower skin
 * Copyright (c) Tom Keffer, Matthew Wall
 * Distributed under terms of GPLv3.  See LICENSE.txt for your rights.
 */

const cookie_prefix = "weewx.power.";
let year_type = get_state('year_type', 'year');

function setup(widgets) {
    // set the state of the history widget
    const id = get_state('history', 'day');
    choose_history(id);
    // if we got a list of widget names, then use it.  otherwise, query the doc
    // for every object with an id of *_widget, and use that as the name list.
    if (!widgets) {
        widgets = [];
        const items = document.getElementsByClassName('widget');
        if (items) {
            for (let i = 0; i < items.length; i++) {
                if (items[i].id) {
                    const widget_name = items[i].id.replace('_widget', '');
                    if (widget_name) {
                        widgets.push(widget_name);
                    }
                }
            }
        }
    }
    // now set the toggle state for each widget based on what the cookies say
    for (let i = 0; i < widgets.length; i++) {
        const state = get_state(widgets[i] + '.state', 'expanded');
        toggle_widget(widgets[i], state);
    }
}

function choose_history(id) {
    choose_div('history', id, ['day', 'week', 'month', 'year']);
    choose_col('hilo', id, ['week', 'month', 'year', 'rainyear']);
    choose_col('totals', id, ['week', 'month', 'year', 'rainyear']);
    choose_rainyear(id);
}

function choose_rainyear(id) {
    if (id === 'year') {
        choose_col('hilo', year_type, ['year', 'rainyear']);
        choose_col('totals', year_type, ['year', 'rainyear']);
    }
}

function toggle_rainyear() {
    if (year_type === 'year') {
        year_type = 'rainyear';
    } else {
        year_type = 'year';
    }
    set_state('year_type', year_type);
    const id = get_active_div('history', ['day', 'week', 'month', 'year'], 'day');
    choose_rainyear(id);
}

function toggle_widget(id, state) {
    const id_elements = document.getElementById(id + '_widget');
    if (id_elements) {
        for (let i = 0; i < id_elements.childNodes.length; i++) {
            if (id_elements.childNodes[i].className === 'widget_contents') {
                if (state === undefined) {
                    // make it the opposite of the current state
                    state = id_elements.childNodes[i].style.display === 'block' ? 'collapsed' : 'expanded';
                }
                id_elements.childNodes[i].style.display = (state === 'expanded') ? 'block' : 'none';
            }
        }
        set_state(id + '.state', state);
    }
}

function choose_col(group, selected_id, all_ids) {
    for (let i = 0; i < all_ids.length; i++) {
        let elements = document.getElementsByClassName(group + '_' + all_ids[i]);
        if (elements) {
            const display = selected_id === all_ids[i] ? '' : 'none';
            for (let j = 0; j < elements.length; j++) {
                elements[j].style.display = display;
            }
        }
    }
}

function choose_div(group, selected_id, all_ids) {
    for (let i = 0; i < all_ids.length; i++) {
        const button = document.getElementById('button_' + group + '_' + all_ids[i]);
        if (button) {
            button.className = (all_ids[i] === selected_id) ? 'button_selected' : 'button';
        }
        const element = document.getElementById(group + '_' + all_ids[i]);
        if (element) {
            element.style.display = (all_ids[i] === selected_id) ? 'block' : 'none';
        }
    }
    set_state(group, selected_id);
}

/* if cookies are disabled, then we must look at page to get state */
function get_active_div(group, all_ids, default_value) {
    let id = default_value;
    for (let i = 0; i < all_ids.length; i++) {
        const button = document.getElementById('button_' + group + '_' + all_ids[i]);
        if (button && button.className === 'button_selected') {
            id = all_ids[i];
        }
    }
    return id;
}

function set_state(name, value, dur) {
    const full_name = cookie_prefix + name;
/*    set_cookie(full_name, value, dur); */
    window.localStorage.setItem(full_name, value);
}

function get_state(name, default_value) {
    const full_name = cookie_prefix + name;
/*    return get_cookie(name, default_value); */
    let value = window.localStorage.getItem(full_name);
    if (value === undefined || value == null) {
        value = default_value;
    }
    return value
}

function set_cookie(name, value, dur) {
    if (!dur) dur = 30;
    const today = new Date();
    let expire = new Date();
    expire.setTime(today.getTime() + 24 * 3600000 * dur);
    document.cookie = name + "=" + encodeURI(value) + ";expires=" + expire.toUTCString();
}

function get_cookie(name, default_value) {
    if (name === "") return default_value;
    const cookie = " " + document.cookie;
    let i = cookie.indexOf(" " + name + "=");
    if (i < 0) i = cookie.indexOf(";" + name + "=");
    if (i < 0) return default_value;
    let j = cookie.indexOf(";", i + 1);
    if (j < 0) j = cookie.length;
    return unescape(cookie.substring(i + name.length + 2, j));
}

function get_parameter(name) {
    const query = window.location.search.substring(1);
    if (query) {
        const vars = query.split("&");
        for (let i = 0; i < vars.length; i++) {
            const pair = vars[i].split("=");
            if (pair[0] === name) {
                return pair[1];
            }
        }
    }
    return false;
}

function load_file(div_id, var_name) {
    let content;
    const file = get_parameter(var_name);
    if (file) {
        content = "Loading " + file;
        let xhr = new XMLHttpRequest();
        xhr.onload = function () {
            let e = document.getElementById(div_id);
            if (e) {
                e.textContent = this.responseText;
            }
        };
        xhr.open('GET', file);
        xhr.send();
    } else {
        content = 'nothing specified';
    }
    let e = document.getElementById(div_id);
    if (e) {
        e.innerHTML = content;
    }
}

function openPOWERFile(date) {
    if (date.match(/^\d\d\d\d/)) {
        window.location = "POWER/POWER-" + date + ".txt";
    }
}

function openTabularFile(date) {
    if (date.match(/^\d\d\d\d/)) {
        window.location = "tabular.html?report=POWER/POWER-" + date + ".txt";
    }
}


install.py
#
#    Copyright (c) 2021-2022 Tom Keffer <tkeffer@gmail.com>
#
#    See the file LICENSE.txt for your full rights.
#
"""Installer for weewx-brultech"""

from distutils.version import StrictVersion
from io import StringIO
import fnmatch
import os

import configobj

import weewx
from weecfg.extension import ExtensionInstaller

REQUIRED_WEEWX = "4.2.0"
if StrictVersion(weewx.__version__) < StrictVersion(REQUIRED_WEEWX):
    raise weewx.UnsupportedFeature("weewx %s or greater is required, found %s"
                                   % (REQUIRED_WEEWX, weewx.__version__))


def loader():
    return WeepwrInstaller()


BRULTECH_DEFAULTS = u"""
[Brultech]

    # See README.md for instructions on how to configure the Brultech devices!!

    # Power is computed as the difference between energy records. How old a 
    # record can be and still be used for this calculation:
    stale = 1800

    # How often to poll the device for data
    poll_interval = 10

    # Max number of channels to emit. Limit is set by hardware (48 for GEM).
    max_channels = 32

    # The type of packet to use. Possible choices are GEMBin48NetTime, GEMBin48Net,
    # or GEMAscii:
    packet_type = GEMBin48NetTime

    # Max number of times to try an I/O operation before declaring an error
    max_tries = 3

    driver = user.brultech

    # The type of connection to use. It should match a section below. 
    # Right now, only 'socket' is supported.
    connection = socket

    # The following is for socket connections: 
    [[socket]]
        host = 192.168.1.104
        port = 8000
        timeout = 20
        # After sending a command, how long to wait before looking for a response    
        send_delay = 0.2

    [[sensor_map]]

[DataBindings]

    [[bt_binding]]
        # The database must match one of the sections in [Databases].
        database = bt_sqlite
        # The name of the table within the database
        table_name = archive
        # The manager handles aggregation of data for historical summaries
        manager = weewx.manager.DaySummaryManager
        # The schema defines the structure of the database.
        # It is *only* used when the database is created.
        schema = user.gem_schema.schema

[Databases]
    [[bt_sqlite]]
        database_name = weepwr.sdb
        database_type = SQLite

[StdReport]
    [[SeasonsPowerReport]]
        skin = SeasonsPower
        enable = True
        data_binding = bt_binding                                    
        [[[Units]]]
            [[[[Groups]]]]
                group_energy2 = kilowatt_hour
            
    [[StandardPowerReport]]
        skin = StandardPower
        enable = False
        data_binding = bt_binding                                    
        [[[Units]]]
            [[[[Groups]]]]
                group_energy2 = kilowatt_hour
            
"""

defaults_dict = configobj.ConfigObj(StringIO(BRULTECH_DEFAULTS), encoding='utf-8')


class WeepwrInstaller(ExtensionInstaller):
    def __init__(self):
        super(WeepwrInstaller, self).__init__(
            version="2.2.0",
            name='weewx-brultech',
            description='Extensions to the weewx weather system for Brultech energy monitors.',
            author="Thomas Keffer",
            author_email="tkeffer@gmail.com",
            config=defaults_dict,
            data_services='user.brultech.BrultechService',
            files=[('bin/user',
                    ['bin/user/gem_schema.py',
                     'bin/user/brultech.py']),
                   ('skins/StandardPower',
                    ['skins/StandardPower/skin.conf',
                     'skins/StandardPower/year.html.tmpl',
                     'skins/StandardPower/weewx.css',
                     'skins/StandardPower/week.html.tmpl',
                     'skins/StandardPower/month.html.tmpl',
                     'skins/StandardPower/index.html.tmpl',
                     'skins/StandardPower/favicon.ico',
                     'skins/StandardPower/backgrounds/band.gif',
                     'skins/StandardPower/Summary/Summary-YYYY.txt.tmpl',
                     'skins/StandardPower/Summary/Summary-YYYY-MM.txt.tmpl']),
                   ('skins/SeasonsPower',
                    ['skins/SeasonsPower/skin.conf',
                     'skins/SeasonsPower/about.inc',
                     'skins/SeasonsPower/identifier.inc',
                     'skins/SeasonsPower/current.inc',
                     'skins/SeasonsPower/index.html.tmpl',
                     'skins/SeasonsPower/hilo.inc',
                     'skins/SeasonsPower/titlebar.inc',
                     'skins/SeasonsPower/power.js',
                     'skins/SeasonsPower/rss.xml.tmpl',
                     'skins/SeasonsPower/statistics.html.tmpl',
                     'skins/SeasonsPower/power.css',
                     'skins/SeasonsPower/statistics.inc',
                     'skins/SeasonsPower/favicon.ico',
                     'skins/SeasonsPower/tabular.html.tmpl',
                     'skins/SeasonsPower/POWER/POWER-YYYY-MM.txt.tmpl',
                     'skins/SeasonsPower/POWER/POWER-YYYY.txt.tmpl',
                     'skins/SeasonsPower/font/Kanit-Bold.ttf',
                     'skins/SeasonsPower/font/OpenSans-Regular.ttf',
                     'skins/SeasonsPower/font/Kanit-Regular.ttf',
                     'skins/SeasonsPower/font/OFL.txt',
                     'skins/SeasonsPower/font/OpenSans.woff2',
                     'skins/SeasonsPower/font/OpenSans.woff',
                     'skins/SeasonsPower/font/license.txt',
                     'skins/SeasonsPower/font/OpenSans-Bold.ttf',
                     'skins/SeasonsPower/lang/en.conf']),
            ]
        )

    def configure(self, engine):
        engine.config_dict['Station']['station_type'] = 'Brultech'
        engine.config_dict['StdArchive']['data_binding'] = 'bt_binding'
        return True
